### DISCUSSION TITLE
Mapping NCBI accessions/GIs to Uniprot IDs produces few hits

### DISCUSSION THREAD

**Discussion by laurahspencer** [RobertsLab/resources#1419] [Q&A]
I have a set of NCBI protein/nucleotide entries in a genome annotation file, and want to ultimately perform GO enrichment analysis.  It seems that the best approach is to map the NCBI entries to Uniprot IDs using this [Uniprot](https://www.uniprot.org/uploadlists/) tool which also retrieves GO IDs.  My issue is that very few of my entries map to Uniprot IDs, for example: 

Using NCBI accession numbers as my input:  
  - 243 out of 13883 RefSeq Protein identifiers were successfully mapped to 244 UniProtKB IDs.  (Out of the remaining ids, 10495 were successfully mapped to the UniParc sequence archive) 
  - 11 out of 722 RefSeq Nucleotide identifiers were successfully mapped to 11 UniProtKB ID.    

Using NCBI GI numbers as my input (which I retrieved using the [Batch Entrez tool](https://www.ncbi.nlm.nih.gov/sites/batchentrez): 
  - 664 out of 14605 GI number identifiers were successfully mapped to 665 UniProtKB IDs in the table below. (Out of the remaining ids, 270 were successfully mapped to the UniParc sequence archive)  

Anyone else run into this issue?  Or has anyone used an enrichment analysis tool that uses NCBI accession numbers as the input? 
FYI I've also tried entering the accession and GI numbers in DAVID, but it doesn't recognize them. As reference, here are a few of the entries:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/LAURA~1.SPE/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/LAURA~1.SPE/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

NCBI_Accession.version | NCBI_Accession | NCBI_GI
-- | -- | --
CAA68079.1 | CAA68079 | 1483311
CAA67357.1 | CAA67357 | 1771447
AAB94634.1 | AAB94634 | 2739396
AAC61691.1 | AAC61691 | 3660667
P81576.1 | P81576 | 5921936
CAC13698.1 | CAC13698 | 14089939
AAK50057.2 | AAK50057 | 14861878
AAK77775.1 | AAK77775 | 15021498
BAB64297.1 | BAB64297 | 15425681

</body>

</html>

NOTE: I have seen [this issue](https://github.com/RobertsLab/resources/issues/1220) and [sam's notebook entry](https://nbviewer.org/github/RobertsLab/code/blob/master/notebooks/sam/20210601_ssal_gff-annotations.ipynb) where he uses a python script to retrieve UniprotIDs, which I haven't tested on my gene sets, however I presume that it would use the same Uniprot database and therefore the same results.

**Comment by sr320** [comment]
If it was me, I would take all the sequences and blast to Uniprot to get associated accessions.

**Reply by laurahspencer** [reply]
I did that too, which I could certainly use, but Giles did a more comprehensive blast approach which resulted in a much higher percentage of annotated genes.

**Reply by sr320** [reply]
I guess what I mean is use your new IDs and batch retrieve fasta, then blast (again)- it also just might be there are limited "hits" but I would not trust unless I did the blast.

**Reply by sr320** [reply]
However that is just doing things in circles, sorry.... If your set on DAVID you can just feed it any and all IDs...

**Reply by laurahspencer** [reply]
I'm definitely not set on DAVID - I was hoping to use GO_MWU, which requires GO terms as input. But I was resorting to DAVID, hoping the IDs I had (or could retrieve) would work in DAVID - they do not.

**Comment by kubu4** [comment]
> I have a set of NCBI protein/nucleotide entries in a genome annotation file

Can you please provide link to file(s)?

> NCBI GI numbers as my input 

Out of curiosity, why are you using these? I'm fairly certain these have been deprecated (years ago) and are not really the standard way to identify NCBI "stuff".

Which UniProt database were you using (reviewed/unreviewd)? Unreviewed should get you more matches.

**Reply by sr320** [reply]
@kubu4 keep in mind her end goal 
> want to ultimately perform GO enrichment analysis

**Reply by kubu4** [reply]
Yep. Need to get GO terms.

But, I'm confused by this, since this addresses what you proposed:

> I did that too, which I could certainly use, but Giles did a more comprehensive blast approach which resulted in a much higher percentage of annotated genes.

@laurahspencer , have you not used the results from these BLASTs?

**Reply by sr320** [reply]
>  have you not used the results from these BLASTs? 

that's the ticket... using these other IDs instead of Uniprot Accessions in DAVID

**Reply by laurahspencer** [reply]
@kubu4 here is the annotation file: [bestblast.bkc_ref.tab](https://raw.githubusercontent.com/laurahspencer/red-king_RNASeq-2022/main/references/bestblast.bkc_ref.tab)

I'm not using all the entries- I filtered for e-value and those with NCBI accession numbers.

**Reply by laurahspencer** [reply]
The annotation file I'm working with is one Giles generated - he blasted against multiple NCBI databases which resulted in ~80% success rate.  While I was waiting for his to finish, I also blasted against the Uniprot database, resulting in far fewer hits.  So, I was proceeding with his results. I have been running into major issues, though, trying to use the NCBI accession numbers (with or without the version number) in any gene enrichment or gene ID conversion tool.  That's why I tried to use the NCBI GI instead, which seems more readily accepted (e.g. you used that in your NCBI -> Uniprot mapping python script).

**Reply by laurahspencer** [reply]
> Which UniProt database were you using (reviewed/unreviewd)? Unreviewed should get you more matches.

That's a great question! I will check that now.  
UPDATE: If I understand your question, I'm using the unreviewed one (UniprotKB).

**Comment by kubu4** [comment]
> NCBI protein/nucleotide entries

Also, which do you have - nucleotides or proteins?

If both, do you get different results when trying one or the other? @yaaminiv had a problem with this when using the Roslin genome (_C.gigas_ ?). You can see some discussions [here](https://github.com/RobertsLab/resources/discussions/1227) and [here](https://github.com/RobertsLab/resources/discussions/1226#discussioncomment-816460). Ultimately, I think she had to use protein accessions to get any matches...

**Reply by kubu4** [reply]
Never mind! I see you tried both!

**Comment by kubu4** [comment]
Having perused your BLAST output table, the primary issue is you have a mix of database IDs. Some are GenBank (e.g. QYJ58449.1), some are RefSeq (e.g. XP_033432496.1). This will make things difficult to deal with when trying to map from one database to another (e.g. the UniProt ID mapping service). The service expects a single IDtype (e.g. GenBank) to mapped to a single database (e.g. UniProt SwissProt). A mix of ID types like you have in your BLAST table cannot be mapped via a "batch" process. You'd have to parse out the different IDs, based on the ID type (which is not indicated in the BLAST table).

To add to this, not all RefSeq accessions are mapped to/from UniProt. It seems like RefSeq mapping criteria is pretty stringent:

https://www.uniprot.org/help/ncbi_mappings

So, with all that in mind, I really think you should take @sr320's approach and perform a BLAST to SwissProt yourself in order to obtain SwissProt accessions. This will then allow you to do a batch submission to UniProt to obtain gene ontology terms.

**Reply by laurahspencer** [reply]
Ah- that makes a lot of sense. I didn't realize that having multiple ID types would mess up conversion - I assumed it would recognize some, and discard others that didn't match the criteria.  Thanks for having a look!

**Reply by sr320** [reply]
I think in early versions of DAVID it could actually use multiple ID types, but now it does not...
